netgrasp/utils/mac.py

```diff
@@ -0,0 +1,255 @@
+        print(wrcfg)
+        return
+    print('source successfully set to "%s"' % source)
+    if source == 'local':
+        print('mac database was last updated: %s' % get_dbupdated_text(updated))
+        if updated == DEFAULT_LOOKUP_UPDATED:
+            print(updated_msg_never)
+        else:
+            print(updated_msg_remind)
+
+
+def update_db(dbfile, lockfile, debugger):
+    print('starting mac database update')
+    try:
+        if TEST_WSDATA:
+            with open(os.path.expanduser(TEST_WSDATA_FILENAME)) as f:
+                wsdata = f.read()
+            wsdata = wsdata.decode('utf-8', errors='replace')
+        else:
+            wsdata = download_wireshark_data(debugger)
+        dbdata = parse_wireshark_data(wsdata)
+        numrecs = len(dbdata)
+        if numrecs:
+            debugger.info('parsed %d records for import' % numrecs)
+        else:
+            raise Exception('parse error: %d records parsed for import' % numrecs)
+        import_data(dbdata, dbfile, lockfile, debugger)
+    except Exception as e:
+        debugger.dump_exception('mac database update failed')
+        print('mac database update failed: %s' % e)
+    else:
+        print(WIRESHARK_ATTRIBUTION)
+        print('mac database update successful')
+
+
+def get_lookup_config(dbfile, lockfile, debugger):
+    from netgrasp.database import database
+
+    try:
+        db = database.Database(dbfile, lockfile, debugger)
+        source = db.get_state(DB_LOOKUP_CONFIG_SOURCE_KEY)
+        updated = db.get_state(DB_LOOKUP_CONFIG_UPDATED_KEY, is_date=True)
+        if source is None or updated is None:
+            raise Exception('config values not found')
+        config = {}
+        config['source'] = source
+        config['updated'] = updated
+    except Exception as e:
+        return False, 'error reading mac lookup config: %s' % e
+
+    return True, config
+
+
+def write_lookup_config(dbfile, lockfile, debugger, source=None, updated=None):
+    from netgrasp.database import database
+
+    source = source or DEFAULT_LOOKUP_SOURCE
+    updated = updated or DEFAULT_LOOKUP_UPDATED
+
+    if source not in LOOKUP_SOURCES:
+        return False, 'invalid mac lookup source "%s", must be one of: %r' % (source, LOOKUP_SOURCES)
+
+    try:
+        db = database.Database(dbfile, lockfile, debugger)
+        db.set_state(DB_LOOKUP_CONFIG_SOURCE_KEY, source)
+        db.set_state(DB_LOOKUP_CONFIG_UPDATED_KEY, updated)
+    except Exception as e:
+        return False, 'error writing mac lookup config: %s' % e
+
+    return True, dict(source=source, updated=updated)
+
+
+def get_dbupdated_text(updated):
+    text = 'never' if updated == DEFAULT_LOOKUP_UPDATED else '%s' % updated
+    return text
+
+
+def import_data(dbdata, dbfile, lockfile, debugger):
+    from netgrasp.database import database
+
+    debugger.info('importing data')
+    db = database.Database(dbfile, lockfile, debugger)
+    db.cursor.execute("DELETE FROM %s" % DB_LOOKUP_TABLE_NAME)
```

Comment:
I'd rather
1) create a new tmp table
2) load the new data there
3) grab lock
4) deleted the old table
5) replace the old table with the new
6) release lock

this minimizes the amount of time things are blocked by the update